mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMemorySearchHighlighter.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.james.mailbox.store.search.SearchHighlighterConfiguration;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.highlight.Formatter;
+import org.apache.lucene.search.highlight.Fragmenter;
+import org.apache.lucene.search.highlight.Highlighter;
+import org.apache.lucene.search.highlight.InvalidTokenOffsetsException;
+import org.apache.lucene.search.highlight.QueryScorer;
+import org.apache.lucene.search.highlight.SimpleFragmenter;
+import org.apache.lucene.search.highlight.SimpleHTMLFormatter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableSet;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class LuceneMemorySearchHighlighter implements SearchHighlighter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneMemorySearchHighlighter.class);
+
+    private final LuceneMessageSearchIndex luceneMessageSearchIndex;
+    private final Analyzer analyzer;
+    private final Formatter formatter;
+    private final Fragmenter fragmenter;
+    private final MessageId.Factory messageIdFactory;
+
+    public LuceneMemorySearchHighlighter(LuceneMessageSearchIndex luceneMessageSearchIndex,
+                                         SearchHighlighterConfiguration searchHighlighterConfiguration,
+                                         MessageId.Factory messageIdFactory) {
+        this.luceneMessageSearchIndex = luceneMessageSearchIndex;
+        this.messageIdFactory = messageIdFactory;
+        this.analyzer = new StandardAnalyzer();
+        this.formatter = new SimpleHTMLFormatter(searchHighlighterConfiguration.preTagFormatter(), searchHighlighterConfiguration.postTagFormatter());
+        this.fragmenter = new SimpleFragmenter(searchHighlighterConfiguration.fragmentSize());
+    }
+
+    @Override
+    public Flux<SearchSnippet> highlightSearch(MultimailboxesSearchQuery expression, MailboxSession session, long limit) {
+        // Todo check if the mailbox is accessible
+
+        SearchQuery searchQuery = expression.getSearchQuery();
+        ImmutableSet<MailboxId> inMailboxes = expression.getInMailboxes();
+        return Mono.fromCallable(() -> luceneMessageSearchIndex.searchDocument(inMailboxes, searchQuery, (int) limit)).flatMapMany(Flux::fromIterable)
+            .map(document -> Throwing.supplier(() -> buildSearchSnippet(document, searchQuery)).get())
+            .doOnError(e -> LOGGER.error("Error while highlighting search", e))
```

Comment:
We are still propagating the error so no need to log it here: it will be logged in double